fms/models/gpt_bigcode.py

```diff
@@ -0,0 +1,324 @@
+        self.config = config
+
+        self.layers = nn.ModuleList(
+            [GPTBigCodeBlock(self.config) for _ in range(self.config.nlayers)]
+        )
+
+        self.embedding = nn.Embedding(
+            self.config.src_vocab_size,
+            self.config.emb_dim,
+            padding_idx=self.config.pad_id,
+        )
+        self.position_embedding = nn.Embedding(self.config.max_pos, self.config.emb_dim)
+        self.register_buffer("pos_id", torch.arange(self.config.max_pos).unsqueeze(0))
+
+        self.dec_norm = nn.LayerNorm(self.config.emb_dim, eps=self.config.ln_eps)
+
+        if self.config.emb_dropout:
+            self.emb_dropout = nn.Dropout(self.config.emb_dropout)
+
+        if self.config.p_dropout:
+            self.dropout = nn.Dropout(self.config.p_dropout)
+
+    def forward(
+        self,
+        x: torch.LongTensor,
+        mask: Optional[torch.Tensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        past_key_value_states: Optional[
+            Tuple[
+                torch.FloatTensor,
+            ]
+        ] = None,
+        use_cache: bool = False,
+        attn_algorithm: Optional[str] = None,
+    ):
+        # Embed the given vocabulary indices using the given attention mask, with pre-/post-norm and dropout as specified
+        # x_in: batch_size x seq_len
+        # mask: batch_size x seq_len x seq_len
+        # bias: nheads x seq_len x seq_len
+
+        qlen = x.size(1)
+        klen = x.size(1)
+
+        if past_key_value_states is None or len(past_key_value_states) == 0:
+            past_key_value_states = [None for _ in range(len(self.layers))]
+
+        # if we are using the cache, the key length needs to be extended with the past keys length
+        if use_cache and past_key_value_states[0] is not None:
+            klen += past_key_value_states[0][0].size(-2)
+
+        # if mask is none, we need to compute mask
+        is_causal_mask = False
+        if mask is None:
+            if x is None:
+                raise ValueError("cannot create a mask when x is None")
+            # we are caching and can assume all 1s in the mask
+            if use_cache and klen != 1 and qlen == 1:
+                # b x h x qlen x kvlen
+                mask = torch.ones(qlen, klen, device=x.device)
+            else:
+                pad_id: int = self.config.pad_id
+                is_pad: torch.BoolTensor = x == pad_id
+                mask: torch.BoolTensor = is_pad.unsqueeze(-1) == is_pad.unsqueeze(-2)
+                mask = mask.tril(diagonal=0)
+
+        x_emb = self.embedding(x)
+
+        if position_ids is None:
+            _position_ids = torch.arange(
+                0, qlen, dtype=torch.long, device=x.device
+            ).repeat(x.size(0), 1)
+            # Compute position_ids based on cache config
+            if use_cache and past_key_value_states[0] is not None:
+                _position_ids += past_key_value_states[0][0].shape[2]
```

Comment:
does this model compile? we noticed at least in some cases we had better luck getting a model to compile when using tensor.size() over tensor.shape

Reply:
I have not tried this yet with compile, I will switch this to size to keep it consistent with how we are doing llama.

Reply:
updated in latest commit